### PR TITLE
configs,ext: Updated the gem5 SST Bridge to use SST 13.0.0

### PR DESCRIPTION
**Pull request by kaustav-goswami**
This change updates the gem5 SST Bridge to use SST 13.0.0. Changes are made to replace SimpleMem class to StandardMem class as SimpleMem will be deprecated in SST 14 and above. In addition, the translator.hh is updated to translate more types of gem5 packets. A new parameter `ports` was added on SST's side when invoking the gem5 component which does not require recompiling the gem5 component whenever a new outgoing bridge is added in a gem5 config.

Change-Id: I45f0013bc35d088df0aa5a71951422cabab4d7f7

### CODE REVIEW

File: configs/example/sst/riscv_fs.py [hunk 1]
```diff
@@ -104,7 +104,7 @@ def createHiFivePlatform(system):
 
     system.platform.pci_host.pio = system.membus.mem_side_ports
 
-    system.platform.rtc = RiscvRTC(frequency=Frequency("100MHz"))
+    system.platform.rtc = RiscvRTC(frequency=Frequency("10MHz"))
```
Comment (BobbyRBruce) [inline configs/example/sst/riscv_fs.py:107]: Why are we doing this?

File: ext/sst/sst/arm_example.py [hunk 2]
```diff
@@ -35,6 +23,18 @@
...
+# All rights reserved.
+#
+# The license below extends only to copyright in the software and shall
+# not be construed as granting a license to any other intellectual
+# property including but not limited to intellectual property relating
+# to a hardware implementation of the functionality of the software
+# licensed hereunder.  You may use the software subject to the license
+# terms below provided that you ensure that this notice is replicated
+# unmodified and in its entirety in all distributions of the software,
+# modified or unmodified, in source code or in binary form.
```
Comment (BobbyRBruce) [inline ext/sst/sst/arm_example.py]: Why did you move this? "The license below extends..." makes no sense if it's at the bottom.

File: ext/sst/sst_responder_subcomponent.cc [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2021 The Regents of the University of California
+// Copyright (c) 2023 The Regents of the University of California
```
Comment (BobbyRBruce) [inline ext/sst/sst_responder_subcomponent.cc]: You've done this all over but you want to do `2021-2023`. What you're saying here is you don't claim copyright for anything prior to 2023.

Fix this in the places you've done this.

**Review by BobbyRBruce — COMMENTED** [review]
It's not a _huge_ problem but this is one of those cases where it's nicer to break up your commits. You made a lot of formatting changes here in addition to just updating the SST bridge. It'd be nice to have these two things segregated so I don't lose focus reviewing. That's not a request for change, just an FYI for the future. You can have more than one commit in a PR.

I trust you've actually run the sst example? I believe that's the only test we have.

**Comment by BobbyRBruce** [comment]
@hnpl : Can you give this a lookover. You're the closest we have to an expect in this.

File: ext/sst/sst/arm_example.py (hunk 2 above)
Comment (kaustav-goswami) [reply ext/sst/sst/arm_example.py]: Oh, I thought the latest commit goes on the top. I'll fix it in the next patch.

File: configs/example/sst/riscv_fs.py (hunk 1 above)
Comment (kaustav-goswami) [reply configs/example/sst/riscv_fs.py:107]: The clocks between the simulators are correctly synchronized. However, when the kernel boots and clocksource changes to the RiscvRTC, the frequency no longer is synchronized, which becomes 10x slower.

File: ext/sst/sst_responder_subcomponent.cc (hunk 3 above)
Comment (kaustav-goswami) [reply ext/sst/sst_responder_subcomponent.cc]: I'll fix this in the next patch.

**Comment by kaustav-goswami** [comment]
> It's not a _huge_ problem but this is one of those cases where it's nicer to break up your commits. You made a lot of formatting changes here in addition to just updating the SST bridge. It'd be nice to have these two things segregated so I don't lose focus reviewing. That's not a request for change, just an FYI for the future. You can have more than one commit in a PR.
> 
> I trust you've actually run the sst example? I believe that's the only test we have.

The reason that I have put this file within this commit as the file is also related to the SST setup that is needed to run the example entire example.

I can actually split this up into two commits. This example has one more correction to be made. I can revert this file and push another change as a separate commit with `configs` tag.

File: ext/sst/INSTALL.md [hunk 4]
```diff
@@ -11,14 +11,14 @@ installed.
 ### Downloading the SST-Core Source Code
 
 ```sh
-wget https://github.com/sstsimulator/sst-core/releases/download/v11.1.0_Final/sstcore-11.1.0.tar.gz
-tar xf sstcore-11.1.0.tar.gz
+wget https://github.com/sstsimulator/sst-core/releases/download/v13.0.0_Final/sstcore-13.0.0.tar.gz
+tar xvf sstcore-13.0.0.tar.gz
```
Comment (BobbyRBruce) [inline ext/sst/INSTALL.md]:
```suggestion
tar xzf sstcore-13.0.0.tar.gz
```

Minor nits here:

1. `v` is just `verbose`. This is an uncesssary addition/
2. I like to be explicit and add `z` which means to also unzip (though it does do this automatically).

File: ext/sst/INSTALL.md [hunk 5]
```diff
@@ -36,14 +36,14 @@ export PATH=$SST_CORE_HOME/bin:$PATH
 ### Downloading the SST-Elements Source Code
 
 ```sh
-wget https://github.com/sstsimulator/sst-elements/releases/download/v11.1.0_Final/sstelements-11.1.0.tar.gz
-tar xf sstelements-11.1.0.tar.gz
+wget https://github.com/sstsimulator/sst-elements/releases/download/v13.0.0_Final/sstelements-13.0.0.tar.gz
+tar xvf sstelements-13.0.0.tar.gz
```
Comment (BobbyRBruce) [inline ext/sst/INSTALL.md]:
```suggestion
tar xzf sstelements-13.0.0.tar.gz
```

File: ext/sst/INSTALL.md [hunk 6]
```diff
@@ -58,24 +58,34 @@ echo "export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:$SST_CORE_HOME/lib/pkgconfig/" >>
 
 ### Building gem5 library
 
-At the root of gem5 folder,
-
+At the root of the gem5 folder, you need to compile gem5 as a library. This
+varies which OS you use. If you're using Linux, then type the following:
```
Comment (BobbyRBruce) [inline ext/sst/INSTALL.md]:
```suggestion
varies  dependent on which OS you are useing If you're using Linux, then execute the following:
```

File: ext/sst/INSTALL.md [hunk 7]
```diff
@@ -58,24 +58,34 @@ echo "export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:$SST_CORE_HOME/lib/pkgconfig/" >>
...
 ```sh
 scons build/RISCV/libgem5_opt.so -j $(nproc) --without-tcmalloc --duplicate-sources
 ```
+In case you're using Mac, then type the following:
```
Comment (BobbyRBruce) [inline ext/sst/INSTALL.md:67]:
```suggestion
If you are using Mac OS X, then execute the following:
```

File: ext/sst/INSTALL.md [hunk 8]
```diff
@@ -58,24 +58,34 @@ echo "export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:$SST_CORE_HOME/lib/pkgconfig/" >>
...
 
-At the root of gem5 folder,
-
+Go to the SST directory in the gem5 repo.
 ```sh
 cd ext/sst
-make
 ```
-
+According to the OS that you're using, you need to rename the `Makefile.xxx` to `Makefile`.
```
Comment (BobbyRBruce) [inline ext/sst/INSTALL.md]:
```suggestion
Depending on your OS, you need to copy the correct `Makefile.xxx` file to `Makefile`.
```

File: ext/sst/INSTALL.md [hunk 9]
```diff
@@ -58,24 +58,34 @@ echo "export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:$SST_CORE_HOME/lib/pkgconfig/" >>
...
+```sh
+cp Makefile.xxx Makefile    # linux or mac
+make -j4
+```
+Change `ARCH=RISCV` to `ARCH=ARM` in the `Makefile` in case you're compiling
+for ARM.
```
Comment (BobbyRBruce) [inline ext/sst/INSTALL.md]:
```suggestion
The make file is hardcoded to RISC-V. IN the case you wish to compile to ARM, edit the Makefile by changing `ARCH=RISCV` to `ARCH=ARM`.
```

File: configs/example/sst/riscv_fs.py (hunk 1 above)
Comment (BobbyRBruce) [reply configs/example/sst/riscv_fs.py:107]: I don't understand this. In this file the clock speed is set via an argument. How can we guarantee these are in-sync?

**Review by BobbyRBruce — CHANGES_REQUESTED** [review]

File: configs/example/sst/riscv_fs.py (hunk 1 above)
Comment (kaustav-goswami) [reply configs/example/sst/riscv_fs.py:107]: I looked at the kernel boot time and the simulator's simulation time. There was a difference of 10x. However, I'd still say that we need better tests to verify this. The kernel switches to the RiscvRTC here:

[    0.007518] clocksource: Switched to clocksource riscv_clocksource

Booting the same kernel in a gem5 only setup and gem5 + SST shows the same 10x difference.

File: ext/sst/INSTALL.md (hunk 4 above)
Comment (kaustav-goswami) [reply ext/sst/INSTALL.md]: Sure, I'll update this.

File: ext/sst/INSTALL.md (hunk 5 above)
Comment (kaustav-goswami) [reply ext/sst/INSTALL.md]: Updated in the next patch.

File: ext/sst/INSTALL.md (hunk 6 above)
Comment (kaustav-goswami) [reply ext/sst/INSTALL.md]: done.

File: configs/example/sst/riscv_fs.py (hunk 1 above)
Comment (powerjg) [reply configs/example/sst/riscv_fs.py:107]: What do you mean "the frequency becomes 10x slower" What frequency? 

You need to justify why this needs to be different for SST compared to the rest of gem5.

@kunpai had to change this 100MHz to get boot to work. See https://gem5-review.googlesource.com/c/public/gem5/+/64013

I wonder if Kunal remembers why he chose 100MHz

Comment (kaustav-goswami) [reply configs/example/sst/riscv_fs.py:107]: 1. The clock read by the kernel is 10x slower if the frequency is set to 100MHz. This time is read from the `printk` statements during the kernel boot.
2. The simulation ends ~300 ms (as per stats.txt and also SST), but the `printk` statements in the kernel print the kernel time ~3 sec when the kernel is simulated until an m5 exit is encountered.
3. I tried using 1MHz, which, in my case works (exists with m5 exit from early kernel space) but the kernel clock becomes 10x faster. The default frequency is 1MHz in the RiscvRTC simobject (`src/dev/riscv/HiFive.py`)
```
    # The default 1MHz setting is taken from SiFive's U54MC
    # core complex. Set to other frequencies if necessary.
```
5. The clocks between gem5 and SST are in sync. I tested this by looking at every memory packet that gem5 is sending to SST, and, whether SST is receiving the same packet at the exact same time.

Comment (kunpai) [reply configs/example/sst/riscv_fs.py:107]: > What do you mean "the frequency becomes 10x slower" What frequency?
> 
> You need to justify why this needs to be different for SST compared to the rest of gem5.
> 
> @kunpai had to change this 100MHz to get boot to work. See https://gem5-review.googlesource.com/c/public/gem5/+/64013
> 
> I wonder if Kunal remembers why he chose 100MHz

I believe I chose 100 MHz because:
1. Our original chosen value of 1 MHz did not complete an FS run even after a week of running.
2. The RISCV Board provided in `src/python/gem5/components/boards/riscv_board.py`, initializes the RTC to be 100 MHz as well [https://github.com/gem5/gem5/blob/48a40cf2f5182a82de360b7efa497d82e06b1631/src/python/gem5/components/boards/riscv_board.py#L115C65-L115C65](https://github.com/gem5/gem5/blob/48a40cf2f5182a82de360b7efa497d82e06b1631/src/python/gem5/components/boards/riscv_board.py#L115C65-L115C65)

Comparing IPC values for 100 MHz vs 1 MHz for SE mode did not give much of a difference if I remember correctly.

Comment (powerjg) [reply configs/example/sst/riscv_fs.py:107]: Why is SST different than without SST? 

Is it wrong throughout all of gem5?

Comment (kaustav-goswami) [reply configs/example/sst/riscv_fs.py:107]: I don't know why this is different in SST. Also, this only happens in RISCV.  I have not extensively tested ARM version of SST, but from what I remember, this issue is absent in the `arm_example.py`.

I'll verify the time reported via `stats.txt` and the `printk` statements for both gem5 and gem5 + SST setup. I'll use a similar system in both cases. I'll get back with the results.

Comment (kaustav-goswami) [reply configs/example/sst/riscv_fs.py:107]: I did an experiment with a minimal system with 1 Timing CPU clocked at 1 GHz and 1 GiB DRAM without caches. I executed `riscv-boot-exit-nodisk` which executes `m5 exit` as the kernel init script. Here are the results: https://docs.google.com/spreadsheets/d/1wT6_dCyDIwVWuKAXJb617DeE8E5rDKZ3wcgOjh1S_l8/edit?usp=sharing

Comment (powerjg) [reply configs/example/sst/riscv_fs.py:107]: Here's the results for those of you who don't want to open a google doc

<html>
<body>
<google-sheets-html-origin><style type="text/css"></style>

Sl. | Experiment | Gem5 | gem5 + SST
-- | -- | -- | --
stats.txt | printk | SST clock | stats.txt | printk
1 | 1MHz | 29.52111 | 2.709169 | 36.4687 | 36.468664 | 3.283364
2 | 10MHz | 30.541303 | 28.106481 | 38.1729 | 38.172885 | 34.528904
3 | 100MHz | 48.81478 | 462.931175 | 75.5352 | 75.535206 | 716.822168


</body>
</html>

So, we should change all of these 100MHz to 10MHz, then :)

Can you create a new PR with that change? You don't have to update this PR.

**Review by powerjg — APPROVED** [review]

**Review by hnpl — APPROVED** [review]
LGTM

**Comment by hnpl** [comment]
This is outside of the scope of this change, but since we updated to SST 13.0.0, we should update the test as well.

Here is the test https://github.com/gem5/gem5/blob/develop/tests/deprecated/nightly.sh#L154. I'm not sure why it says deprecated but not ported to supposedly new tests.

Here is the Dockerfile used for testing https://github.com/gem5/gem5/tree/develop/util/dockerfiles/sst-11.1.0.

File: configs/example/sst/riscv_fs.py (hunk 1 above)
Comment (kaustav-goswami) [reply configs/example/sst/riscv_fs.py:107]: Okay, before opening that PR, I just want to test 100MHz with the RISCV board just to be double sure. I'll update you soon with the results.

**Review by BobbyRBruce — APPROVED** [review]

**Comment by powerjg** [comment]
Hmm... is this going to break the daily tests?

See https://github.com/gem5/gem5/blob/stable/.github/workflows/daily-tests.yaml#L214 and #436
